controllers/hardwareclassification_controller.go

```diff
@@ -17,36 +14,222 @@ package controllers
+
+				introspectionDetails["CPU"] = host.Status.HardwareDetails.CPU
+			}
+
+			if extractedProfile.Disk != (hwcc.Disk{}) {
+				if extractedProfile.Disk.MinimumCount < 0 || extractedProfile.Disk.MaximumCount < 0 {
+					err = errors.New("enter valid Disk Count")
+					break
+				}
+
+				if extractedProfile.Disk.MinimumIndividualSizeGB < 0 || extractedProfile.Disk.MaximumIndividualSizeGB < 0 {
+					err = errors.New("enter valid Disk Size in GB")
+					break
+				}
+				introspectionDetails["Disk"] = host.Status.HardwareDetails.Storage
+			}
+
+			if extractedProfile.NIC != (hwcc.NIC{}) {
+				if extractedProfile.NIC.MinimumCount > 0 || extractedProfile.NIC.MaximumCount > 0 {
+					introspectionDetails["NIC"] = host.Status.HardwareDetails.NIC
+				} else {
+					err = errors.New("enter valid NICS Count")
+					break
+				}
+			}
+
+			if extractedProfile.RAM != (hwcc.RAM{}) {
+				if extractedProfile.RAM.MinimumSizeGB > 0 || extractedProfile.RAM.MaximumSizeGB > 0 {
+					introspectionDetails["RAMMebibytes"] = host.Status.HardwareDetails.RAMMebibytes
+				} else {
+					err = errors.New("enter valid RAM size in GB")
+					break
+				}
+			}
+
+			if len(introspectionDetails) > 0 {
+				extractedHardwareDetails[host.ObjectMeta.Name] = introspectionDetails
+			}
+		}
+
+	}
+	if err != nil {
+		return extractedHardwareDetails, err
+	}
+
+	return extractedHardwareDetails, nil
+}
+
+// SetupWithManager will add watches for this controller
 func (r *HardwareClassificationReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
-		For(&metal3iov1alpha1.HardwareClassification{}).
+		For(&hwcc.HardwareClassification{}).
+		Watches(
+			&source.Kind{Type: &bmh.BareMetalHost{}},
+			&handler.EnqueueRequestsFromMapFunc{
+				ToRequests: handler.ToRequestsFunc(r.BareMetalHostToHardwareClassification),
+			},
+		).
 		Complete(r)
 }
+
+// BareMetalHostToHardwareClassification will return a reconcile request for a
+// HardwareClassification if the event is for a BareMetalHost.
+func (r *HardwareClassificationReconciler) BareMetalHostToHardwareClassification(obj handler.MapObject) []ctrl.Request {
+	var result []ctrl.Request
+
+	if len(validHostList) == 0 {
+		validHostList = fetchBmhHostList(context.Background(), r, "metal3")
+	}
+
+	if host, ok := obj.Object.(*bmh.BareMetalHost); ok {
+
+		// If host found in validHostList and current provisioining state
+		// is not ready then remove host from validHostList. Else if host
+		// not found in validHostList and current provisioning state is ready
+		// then append it to validHostList.
+		if checkValidHost[host.ObjectMeta.Name] && host.Status.Provisioning.State != "ready" {
+			for i, validHost := range validHostList {
+				if validHost.ObjectMeta.Name == host.ObjectMeta.Name {
+					validHostList = append(validHostList[:i], validHostList[i+1:]...)
+					checkValidHost[validHost.ObjectMeta.Name] = false
+					result = append(result, ctrl.Request{NamespacedName: name})
```

Comment:
What is in name here? It's not set inside this function.

Reply:
This was declared globally earlier but now it is removed. This is no more required.
Addressed in new commit: c239dc6